### ISSUE TITLE
Fixity Auditing in Islandora

### ISSUE THREAD

**Issue by kstapelfeldt** [Islandora/documentation#1931]
| Title (Goal)  | Fixity Auditing in Islandora                 |
--------------- |------------------------------------                           |
| Primary Actor | Collection Manager   |
| Scope         | Preservation    |
| Story         | As a collection manager I wish to be able to perform fixity audits on objects in the repository and see reports pertaining to the health of objects in my repository.

**Comment by rosiel** [comment]
Is this the same as "As a collection manager, I wish to generate checksums for the media in my repository."?

**Comment by kstapelfeldt** [comment]
They could be added together, but generating checksums would be discrete from checking checksums and seeing reports about it, I think?

**Comment by rosiel** [comment]
Possibly. Do you think there's a use case for "collection managers" to generate checksums that nothing happens with? 

This was from:
- [Feature Parity](https://docs.google.com/spreadsheets/d/1jE19AZgwceTo9-eccCRfseiO-4q4iVGPsMWxC5JZrL4/edit#gid=960987100), line 84, phrased as "Enables checksums on datastreams". 
- ICG Report, "Include and document digital preservation features in Islandora."

It sounds like this is part of a larger, "Preservation" interest/suite of features. I wonder if it'd be worth a dedicated interest group (or, "[Initiative](https://www.drupal.org/community-initiatives)" which seem to be a bit different from our version of IGs)?

**Comment by mjordan** [comment]
Fixity auditing it verifying a file's checksum periodically over time to ensure that the file has not changed inadvertently. So generating checksums is not the same thing as fixity auditing, it's a precondition for fixity auditing.

**Comment by kstapelfeldt** [comment]
What Mark said. I think the use case would be that I have something else that checks the checksums. They could go together. 

I think an interest group could be a great idea, but we know about immediate gaps in the toolset, so I think we can proceed to build things like checksum generation, and fixity checking and reporting (and ostensibly, a lot already exists in @mjordan work that could be embraced more officially)

**Comment by mjordan** [comment]
I am happy to contribute Riprap and Islandora Riprap Integration to the IF (and remain its maintainer). It currently generates  checksums, performs the periodic verification of checksums, and reports checksum mismatches found during that periodic verification.

FYI Islandora already requires the Filehash contrib module, which generates checksums. AFAIK it doesn't do anything specific with them other than store them in a db table.

**Comment by mjordan** [comment]
@kstapelfeldt we have a Preservation IG but it has been inactive for some time. Could revive it.

**Comment by kstapelfeldt** [comment]
I think that could be a good longer-term goal for guiding preservation work in Islandora, so long as there is sufficient interest. For me, RipRap would be an excellent addition to the official repositories of Islandora.

**Comment by DonRichards** [comment]
@mjordan Is it possible to run RipRap without Fedora? Like storing a hash into a field or something :shrug:

**Comment by kstapelfeldt** [comment]
I'm not the tagged person, but FWIW we're running RipRap without Fedora. It's definitely possible and Mjordan wrote this so it will work with Drupal.

**Comment by mjordan** [comment]
@kstapelfeldt is correct. You need to use the PluginFetchDigestFromDrupal plugin instead of the PluginFetchDigestFromFedoraAPI plugin.

**Comment by rosiel** [comment]
The Drupal Filehash module is already part of islandora, and we could improve how we use it!  
* available since filehash 2.0, we can keep an "original" checksum?
* integrate with a local drupal checksum checker? Could it be a view?
* Current behaviour of islandora derivatives "replaces" files on disk, which would cause checksum check errors!
* Discuss best practices around duplicate checking
* Discuss best practices around hashing algorithms
* Make our hardcoded sha-1 check not rely on by luck having the sha-1 algorithm selected.

**Comment by dannylamb** [comment]
Yep, i think i've turned the 'original checksum' thing on a site or two.  We're totally already keeping hashes of everything in the DB on drupal side.  There's a lot to expand on there.  Exciting!

**Comment by rosiel** [comment]
(Tech call) DGI has also released a Fixity Check module: https://github.com/discoverygarden/dgi_fixity
